test/src/userobjects/ExampleShapeElementUserObject.C

```diff
@@ -0,0 +1,103 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "ExampleShapeElementUserObject.h"
+#include "libmesh/quadrature.h"
+
+template<>
+InputParameters validParams<ExampleShapeElementUserObject>()
+{
+  InputParameters params = validParams<ShapeElementUserObject>();
+  params.addRequiredCoupledVar("u", "first coupled variable");
+  params.addRequiredCoupledVar("v", "second coupled variable");
+  return params;
+}
+
+ExampleShapeElementUserObject::ExampleShapeElementUserObject(const InputParameters & parameters) :
+    ShapeElementUserObject(parameters),
+    _u_value(coupledValue("u")),
+    _u_var(coupled("u")),
+    _v_value(coupledValue("v")),
+    _v_var(coupled("v"))
+{
+}
+
+void
+ExampleShapeElementUserObject::initialize()
+{
+  _integral = 0.0;
+
+  // Jacobian term storage is up to the user. One option is using an std::vector
+  // We resize it to the total number of DOFs in the system and zero it out.
+  // WARNING: this can be large number (smart sparse storage could be a future improvement)
+  // if (_currently_computing_jacobian)
+    _jacobian_storage.assign(_subproblem.es().n_dofs(), 0.0);
```

Comment:
@friedmud we've already thought about this and maybe @SudiptaBiswas wants to comment, too. There are imaginable usecases where the Jacobian contribution is largely zero, except for a few very localized areas (however the localized areas affect large areas of the domain - but again not all DOFs). I think our phasefield contact (rigid body) problem falls into this category. The contact regions are where we have interface overlap, and the affected areas are areas where a specific order parameter is nonzero. I believe we can get quite a bit f sparsity out of those cases. 

But if your Jacobian is dense we need a better idea (maybe develop specific preconditioning methods?)